### PR TITLE
contrib: add NVMeoF support

### PR DESCRIPTION
**Pull request by qwren**
This patch will add NVMeoF support.

The NVMe over Fabrics (NVMeOF) is an emerging technology, and it gives
data centers unprecedented access to NVMe SSD storage. Like iSCSI, NVMeОF
is a network protocol used to communicate between a host and a storage
system over a network (aka fabric).

Signed-off-by: Shixin Yang <shixin.yang@intel.com>
Signed-off-by: Wenjie Liu <wenjie.liu@intel.com>
Signed-off-by: Qiaowei Ren <qiaowei.ren@intel.com>

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: contrib/connector/nvmeof/nvmeof_helper.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+	for _, l := range lines {
+		if strings.Contains(l, "UUID:") {
+			tmp := iniNvmePrefix + strings.Split(l, ":")[1]
+			nqns = append(nqns, tmp)
+			log.Printf("Found the following nqns: %s", nqns)
+			return nqns, nil
+		}
+	}
+	log.Printf("Can not find any nqn initiator")
+	return nqns, errors.New("Can not find any nqn initiator")
```
Comment (leonwanghui) [inline contrib/connector/nvmeof/nvmeof_helper.go]: The begin letter of error message should be lower case.

File: contrib/connector/nvmeof/nvmeof_helper.go [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+func getInitiatorInfo() (string, error) {
+
+	initiators, err := GetInitiator()
+	if err != nil {
+		return "", err
+	}
+
+	if len(initiators) == 0 {
+		return "", errors.New("No nqn found")
```
Comment (leonwanghui) [inline contrib/connector/nvmeof/nvmeof_helper.go]: Same here

File: contrib/drivers/lvm/lvm.go [hunk 3]
```diff
@@ -249,23 +252,28 @@ func (d *Driver) InitializeConnection(opt *pb.CreateVolumeAttachmentOpts) (*mode
 	if d.conf.EnableChapAuth {
 		chapAuth = []string{utils.RandSeqWithAlnum(20), utils.RandSeqWithAlnum(16)}
 	}
-	t := targets.NewTarget(d.conf.TgtBindIp, d.conf.TgtConfDir)
+
+	// create target according to the pool's access protocol
+	accPro := opt.AccessProtocol
+		log.Info("accpro:", accPro)
```
Comment (leonwanghui) [inline contrib/drivers/lvm/lvm.go]: Please use `go fmt` to update the code style

File: contrib/drivers/lvm/lvm.go [hunk 4]
```diff
@@ -249,23 +252,28 @@ func (d *Driver) InitializeConnection(opt *pb.CreateVolumeAttachmentOpts) (*mode
...
 	}, nil
 }
 
 func (d *Driver) TerminateConnection(opt *pb.DeleteVolumeAttachmentOpts) error {
-	t := targets.NewTarget(d.conf.TgtBindIp, d.conf.TgtConfDir)
+	accPro := opt.AccessProtocol
+	t := targets.NewTarget(d.conf.TgtBindIp, d.conf.TgtConfDir, accPro)
 	if err := t.RemoveExport(opt.GetVolumeId()); err != nil {
-		log.Error("Failed to initialize connection of logic volume:", err)
+		log.Error("Failed to termiate  connection of logic volume:", err)
```
Comment (leonwanghui) [inline contrib/drivers/lvm/lvm.go]: Same comment here

File: contrib/drivers/lvm/targets/nvmeof.go [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+	_, err = t.execCmd("ln", "-s", sysdir, portssub)
+	if err != nil {
+		log.Errorf("Fail to create link")
+		return err
+	}
+
+	// check
+	info, err := t.execCmd("dmesg", `|grep "enabling port"`)
+	if err != nil || info == "" {
+		log.Errorf("NVMe target is not listening on the port")
```
Comment (leonwanghui) [inline contrib/drivers/lvm/targets/nvmeof.go]: Same here

File: script/devsds/lib/lvm.sh [hunk 6]
```diff
@@ -23,20 +23,42 @@ set +o xtrace
 # Name of the lvm volume groups to use/create for iscsi volumes
 VOLUME_GROUP_NAME=${VOLUME_GROUP_NAME:-opensds-volumes}
 DEFAULT_VOLUME_GROUP_NAME=$VOLUME_GROUP_NAME-default
+#Name of lvm nvme volume group to use/create for nvme volumes
```
Comment (leonwanghui) [inline script/devsds/lib/lvm.sh]: Please add a blank space here

**Comment by xing-yang** [comment]
@qwren Thanks for contributing the NVMeoF driver?  Can you submit your design specs to https://github.com/opensds/design-specs repo?  I believe you have a draft already.  Thanks.

**Comment by qwren** [comment]
> @qwren Thanks for contributing the NVMeoF driver? Can you submit your design specs to https://github.com/opensds/design-specs repo? I believe you have a draft already. Thanks.

Sure, I am about to upstream that doc.

File: contrib/connector/nvmeof/nvmeof_helper.go (hunk 1 above)
Comment (qwren) [reply contrib/connector/nvmeof/nvmeof_helper.go]: You mean that the begin letter of all log information, including error, infor, ...,  should be lower case, right?

Comment (leonwanghui) [reply contrib/connector/nvmeof/nvmeof_helper.go]: Yes

File: contrib/drivers/lvm/targets/targets.go [hunk 7]
```diff
@@ -26,10 +29,18 @@ type Target interface {
 	RemoveExport(volId string) error
 }
 
-// NewTarget method creates a new iscsi target.
-func NewTarget(bip string, tgtConfDir string) Target {
-	return &iscsiTarget{
-		ISCSITarget: NewISCSITarget(bip, tgtConfDir),
+// NewTarget method creates a new target based on its type.
+func NewTarget(bip string, tgtConfDir string, access string) Target {
+	if access == iscsiAccess {
```
Comment (leonwanghui) [inline contrib/drivers/lvm/targets/targets.go]: I think using `switch` could be better.

**Comment by leonwanghui** [comment]
Also please rebase the code

**Comment by wisererik** [comment]
LGTM

**Pull request merged**